### PR TITLE
SVM Analysis using Literate

### PR DESCRIPTION
**Pull request by be-apt**

### CODE REVIEW

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 1]
```diff
@@ -0,0 +1,248 @@
+# # Sensitivity Analysis of SVM
+
+#md # [![](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white)](@__REPO_ROOT_URL__/examples/sensitivity-analysis-svm.jl)
+#md # [![](https://img.shields.io/badge/show-nbviewer-579ACA.svg)](@__NBVIEWER_ROOT_URL__/generated/sensitivity-analysis-svm.ipynb)
+
+
+# This notebook illustrates sensitivity analysis of data points in an [Support Vector Machine](https://en.wikipedia.org/wiki/Support-vector_machine) (inspired from [@matbesancon](http://github.com/matbesancon)'s [SimpleSVMs](http://github.com/matbesancon/SimpleSVMs.jl).)
```
Comment (blegat) [inline docs/src/examples/sensitivity-analysis-svm.jl:7]: an -> a

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+# For reference, Section 10.1 of https://online.stat.psu.edu/stat508/book/export/html/792 gives an intuitive explanation of what does it means to have a sensitive hyperplane or data point. The general form of the SVM training problem is given below (without regularization):
```
Comment (blegat) [inline docs/src/examples/sensitivity-analysis-svm.jl:9]: "what does it means" - "what it means"

**Review by blegat — COMMENTED** [review]
What's the reason we use MOI and not JuMP in the example ?

**Comment by joaquimg** [comment]
Maybe we can even have both...

**Comment by be-apt** [comment]
> What's the reason we use MOI and not JuMP in the example ?

will update the example using JuMP. that way it'll be easier to understand too

**Comment by blegat** [comment]
> Maybe we can even have both...

Yes, and the difference between the two can be quite small, see https://github.com/jump-dev/MathOptInterface.jl/issues/1472#issuecomment-881150059

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+# Add the constraints.
+
+@constraint(
+    model, 
+    1.0*l ∈ MOI.Nonnegatives(N),
+)
+@constraint(
+    model, 
+    cons, 
+    y.*(X*w .+ b) + l.-1 ∈ MOI.Nonnegatives(N),
```
Comment (blegat) [inline docs/src/examples/sensitivity-analysis-svm.jl:61]: Use spaces around binary operators

File: .gitignore [hunk 4]
```diff
@@ -3,6 +3,7 @@
 *.jl.mem
 docs/build/
 docs/site/
+docs/src/examples/*.md
```
Comment (joaquimg) [inline .gitignore:6]: why ignore these?

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+# ## Define and solve the SVM
+
+# Import the libraries.
+
+
+using SCS, DiffOpt, LinearAlgebra, JuMP
+import Random, Plots
+
+
+# Construct separatable, non-trivial data points.
```
Comment (joaquimg) [inline docs/src/examples/sensitivity-analysis-svm.jl:33]: separable?

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+N = 100
+D = 2
+Random.seed!(62)
+X = vcat(randn(N ÷ 2, D), randn(N ÷ 2, D) .+ [4.0, 1.5]')
+y = append!(ones(N ÷ 2), -ones(N ÷ 2));
+
+
+# Let's define the model
```
Comment (joaquimg) [inline docs/src/examples/sensitivity-analysis-svm.jl:42]: Let's initialize a special model that  can understand sensitivities

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+# Add the constraints.
+
+@constraint(
+    model, 
+    1.0 * l ∈ MOI.Nonnegatives(N),
+)
+@constraint(
+    model, 
+    cons, 
+    y .* (X * w .+ b) + l .- 1 ∈ MOI.Nonnegatives(N),
```
Comment (joaquimg) [inline docs/src/examples/sensitivity-analysis-svm.jl:61]: y .* (X * w .+ b) + l .- 1 >= 0 ?

File: docs/src/examples/sensitivity-analysis-svm.jl [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+@variable(model, l[1:N])
+@variable(model, w[1:D])
+@variable(model, b);
+
+# Add the constraints.
+
+@constraint(
+    model, 
+    1.0 * l ∈ MOI.Nonnegatives(N),
```
Comment (joaquimg) [inline docs/src/examples/sensitivity-analysis-svm.jl:56]: `l .>= 0` ?

**Comment by joaquimg** [comment]
Looking good!

File: .gitignore (hunk 4 above)
Comment (be-apt) [reply .gitignore:6]: these are generated by literate right? the source literate file and markdown files in the same folder might be redundant
if you want, we can keep them in the docs folder - that way apart from https://jump.dev/DiffOpt.jl/dev, one can access docs here too

File: docs/src/examples/sensitivity-analysis-svm.jl (hunk 5 above)
Comment (be-apt) [reply docs/src/examples/sensitivity-analysis-svm.jl:33]: changed

File: docs/src/examples/sensitivity-analysis-svm.jl (hunk 7 above)
Comment (be-apt) [reply docs/src/examples/sensitivity-analysis-svm.jl:61]: i get
```
MethodError: no method matching set(::Model, ::DiffOpt.ForwardInConstraint, ::Vector{ConstraintRef{Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.GreaterThan{Float64}}, ScalarShape}}, ::MathOptInterface.VectorAffineFunction{Float64})
```

**Comment by be-apt** [comment]
any changes on this?

File: docs/src/examples/sensitivity-analysis-svm.jl (hunk 7 above)
Comment (blegat) [reply docs/src/examples/sensitivity-analysis-svm.jl:61]: You should use broadcast: `MOI.set.`

**Comment by blegat** [comment]
See https://github.com/jheinen/GR.jl/issues/422 for how to fix the docs build

**Comment by be-apt** [comment]
@matbesancon any idea why this might be failing
```
Differentiating simple PSD back: Error During Test at /home/runner/work/DiffOpt.jl/DiffOpt.jl/test/conic_backward.jl:10
  Got exception outside of a @test
  MethodError: *(::Adjoint{Float64, Vector{Float64}}, ::BlockDiagonals.BlockDiagonal{Float64, Matrix{Float64}}) is ambiguous. Candidates:
    *(M::AbstractMatrix, B::BlockDiagonals.BlockDiagonal{T, V} where V<:AbstractMatrix{T}) where T in BlockDiagonals at /home/runner/.julia/packages/BlockDiagonals/Nvfvo/src/base_maths.jl:108
    *(x::Adjoint{T, <:AbstractVector} where T, A::AbstractMatrix) in LinearAlgebra at /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.8/LinearAlgebra/src/matmul.jl:113
  Possible fix, define
    *(::Adjoint{T, <:AbstractVector} where T, ::BlockDiagonals.BlockDiagonal{T, V} where V<:AbstractMatrix{T}) where T
```

thought https://github.com/invenia/BlockDiagonals.jl/pull/55 fixed it

**Comment by matbesancon** [comment]
I suspect there was a dependency blocking us, the BlockDiagonals version used was 0.1.6 which might be from before the fix

**Comment by matbesancon** [comment]
@be-apt this is fixed, was probably one dependency blocking. Could you check why the docs are failing and whether the change in the other PR would fix it? If so we should probably have a PR fixing the docs action first, and then see to merge this

**Comment by matbesancon** [comment]
I'm still not sure what is wrong with the documentation step here, one plot seems not to be produced, causing an error

**Comment by matbesancon** [comment]
@odow were there some gotchas when you switch the documentation to Literate for JuMP?
This seems to build the markdown fine but errors when making the docs themselves

**Comment by odow** [comment]
Not sure. This looks like it's a problem with the plotting, not Documenter or Literate.

**Comment by blegat** [comment]
I had the same error on https://github.com/JuliaPolyhedra/Polyhedra.jl/pull/265 and fixed it in https://github.com/JuliaPolyhedra/Polyhedra.jl/pull/265/commits/5db5c2d80c92632be371b172fa99542c95969f2d.
See https://github.com/jheinen/GR.jl/issues/422#issuecomment-888953237

**Comment by matbesancon** [comment]
perfect thanks @blegat

**Pull request merged**